### ISSUE TITLE
TouchEvent is being dispatched for left-click on Mac OS target

### ISSUE THREAD

**Issue by seivan** [cocos2d/cocos2d-x#13671]
Right now I am getting a touch event for left-click on a mouse when running the game on Mac OS.

This creates a problem as I am getting both a mouseEvent as well as a touchEvent.
I could add a platform-macro to ignore touch on Desktop targets, but that would break for Windows that uses both Mouse and Touch as it only has _one_ target for Phone/Tablet and Desktop.

This is the code responsible for it. 
https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/platform/desktop/CCGLViewImpl-desktop.cpp#L606-L628

Is this a bug or intended feature?

**Comment by pandamicro** [comment]
Yes, this is intended, so that we can always use touch events on both desktop and mobile platform.

If you want to use mouse event instead of touch event on Mac, you can use : 

```
Application::Platform platform = Application::getInstance()->getTargetPlatform();
if (platform == ApplicationProtocol::Platform::OS_MAC) {
    // Add mouse event listener
}
else {
    // Add touch event listener
}
```

**Comment by seivan** [comment]
@pandamicro But if you right-click you get no touch event. So you have to implement a mouse-event. It's seems counter-intuitive because with a mouse-event and a touch event you will get _two_ events with for left-click.

I will probably use the code snippet provided (thank you!) But I think it's not something you should know when encountering this. 

Does only the Mac target give left-click as touch or also Windows and Linux?

**Comment by pandamicro** [comment]
Yes, if you want to occupy both left and right mouse event, you should only listen to mouse event.
The behavior is the same on Mac, Windows and Linux

**Comment by seivan** [comment]
How about a crossplatform game on Windows and Windows phone.
How do I say do touch on Windows Phone and Mouse Windows since they both share the same target?

**Comment by stevetranby** [comment]
Use the platform defines. #ifdef PLATFORM_WIN32 or PLATFORM_IOS, etc
(the actual macro name may be slightly different)

**Comment by seivan** [comment]
I think it's a poor surprise that shouldn't be there at all. 
If you want touch, use touch, if you want mouse then use mouse. 
The API should never surprise you.

If you're cross plattform you _still_ have to implement both, and now suddenly your left-clicks will trigger twice on desktops.

@stevetranby But Windows share the same Macro for phone/tablets and computers beyond  Win 8. 

Why can't touch be touch and clicks be clicks. Especially since touch doesn't handle other mouse buttons.

**Comment by stevetranby** [comment]
We use touch events for all of our left-click. Then we use Mouse event listener for right-click and scroll wheel. I know we struggled to figure out how we should tackle it, but I do realize it's not perfect.

**Comment by seivan** [comment]
Except your left-click also triggers on Mouse Event listener, which is what I am saying.
Meaning it triggers a duplicate. 

Obviously you can solve that by ignoring left clicks on mouse events. 
But that's not _obvious_ to begin with, making the API poorly designed as it should be an obvious decision.

**Comment by stevetranby** [comment]
I don't know what the solution is, but if you develop your app for touch it is nice to be able to just have it run correctly on a platform using a mouse instead of touch. So for the WinRT platform allowing a person to use a mouse or touch to cause the same "touch" events seems like an obvious basic solution.

In other words 75%+ games (guess) only ever need to worry about using the touch events.

I think that is the design and I don't think it's necessarily poor because of this. Maybe there should be a way to easily toggle or set a flag that makes touches and clicks separate as you'd like.

**Comment by seivan** [comment]
Yeah, the easy way you're thinking of would be have a separate listener for mouse and one for touch. Which we currently have. Except for one tidbit; duplicate events are fired for the left-click. 

You will have duplicate events on Windows that have no separate targets(?)

I am not sure what you're missing or what I am being ambiguous about, this is essentially a bug on Windows and a poor API design on every other target.

A decent API would not surprise you like that, I had to debug the internals to find that it's dispatching two events on click, one for touch and one for click. I thought it was a bug until you told me it's by design. 

**This breaks on Windows as it will fire duplicate events and there is no way to check if it's a Phone/Surface/Desktop**  - correct me if I am wrong, I haven't found anything to make that distinction on > Win 8

**Comment by stevetranby** [comment]
I don't understand what you mean? If you only need to support left-clicking (not right or middle clicks) then just use Touch only for everything. Simple. One event listener. No confusion. Everything just works.
(edit: in other words, don't even bother creating a mouse listener)

Is your issue that you want to support right-clicking?

**Comment by seivan** [comment]
I am saying it's a bug  to dispatch two events when you should be dispatching one.

You dispatch _two_ events 
1) one for touch 
2) one for mouse 
on the _single_ event of a left-click.

And there is _no way(?)_ to add a macro to avoid it on Windows, or to differentiate between Touch and Mouse type targets.
Which makes sense because Windows accepts both mouse and touch actions, which is why I suspect you only have _ONE_ target for both mobile/surface and computers.

This breaks the logic, and thus a bug, e.g my mario will jump twice on click. 

Not to mention poor/surprising API design that touch events will be dispatched on a click happening.

**Comment by stevetranby** [comment]
I'll agree that the API could be improved, even if I think it's easy for us to work with for our combined mobile+desktop game needs (touch + right-mouse + scroll wheel + keyboard). It would be nice if there was an API to get available input devices and which or if they're connected.

**Comment by pandamicro** [comment]
The original thoughts is just as @stevetranby said, we want our engine to be friendly with our targeted users.

Two facts:
- Cocos2d-x have always been mostly popular for mobile games on iOS & Android.
- These developers develop mostly on windows and Mac OS.

So the current implementation is for simplify development on Windows/Mac for iOS/Android, I agree with @seivan that the API doesn't suggest what it really do in this case, but we don't want our user to code every time two listeners (one for mouse on desktop, another for touch on mobile).

In conclusion, it's not a good API design, but a friendly one for most of our users, we don't have plan to modify the current behavior to avoid breaking backward compatibility, but maybe it's a good idea to permit enable/disable it

**Comment by seivan** [comment]
@pandamicro How can it be friendly if it's buggy for Windows users?

I feel like I am repeating myself, there are two events being sent.
If I shoot with left-click then my character will try to shoot twice.
This is a bug foremost, and unexpected API (bad design) second.

I know you wanted to make it easy for Mouse users by just having 1 event listener for both, but that shit breaks down when you want to code for middle-click and right-click
Not to mention it completely breaks down for Windows targets.

Unless there is a way to separate Windows Mobile/Surface from Windows Desktop, the TouchEvent from mouse clicks should be removed.

The user wants Mouse-event, implement Mouse-listener <- you still have to do that for all the other mouse stuff necessary.

However I agree with @stevetranby that in a perfect world the interface would be agnostic. In fact, that's what Apple does in SpriteKit.

If such an pull request was accepted (it would be a big change) I would actually prefer that, and even do the necessary changes myself.

(From Apple)

``` swift
 func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent)
  func touchesMoved(touches: Set<UITouch>, withEvent event: UIEvent){}
  func touchesEnded(touches: Set<UITouch>, withEvent event: UIEvent) {}
  func touchesCancelled(touches: Set<UITouch>, withEvent event: UIEvent) {}
```

`UIEvent` contains details if it's a left-click or a long-touch.
while NSSet contains 1 or more touches (depending on multitouch or not activate)

This would be agnostic.

**Comment by stevetranby** [comment]
We have Win32 and Windows Phone working with the following. I get it's not a perfect API, but it is what it is. If you hate it, write your own GLFW Input wrapper or pull SFML's or SDL's input sub-system into your project and don't use the cocos2d one.

Thanks for the info that Apple provides left/right in event details. That will help them in the future when the core team (I'm not one of them) decides how to best improve the input APIs for v4 and beyond. 

``` cpp
// SetupInput() method
auto listener1 = EventListenerTouchAllAtOnce::create();
listener1->onTouchesBegan = [this](const std::vector<Touch*>& pTouches, Event *pEvent) {
    // code for left-click or touch
}
listener1->onTouchesMoved = [this](const std::vector<Touch*>& pTouches, Event *pEvent) {..}
listener1->onTouchesEnded = [this](const std::vector<Touch*>& pTouches, Event *pEvent) {..}
listener1->onTouchesCancelled = [this](const std::vector<Touch*>& pTouches, Event *pEvent) {..}  
getEventDispatcher()->addEventListenerWithSceneGraphPriority(listener1, this);

// for right-click and wheel
auto listener0 = EventListenerMouse::create();
listener0->onMouseDown = [this,screenRect](Event* event) {
    auto mouseEvent = (EventMouse*)event;
    if(mouseEvent->getMouseButton() == MOUSE_BUTTON_LEFT)
    {
        // handled by touch events
    }
    if(mouseEvent->getMouseButton() == MOUSE_BUTTON_RIGHT) {
        // code here
    }
};
listener0->onMouseMove = [this,screenRect](Event* event) {
    auto mouseEvent = static_cast<EventMouse*>(event);
    if(mouseEvent->getMouseButton() == MOUSE_BUTTON_RIGHT) {..}
};
listener0->onMouseUp = [this,screenRect](Event* event) {
    auto mouseEvent = static_cast<EventMouse*>(event);
    if(mouseEvent->getMouseButton() == MOUSE_BUTTON_RIGHT) {..}
};
listener0->onMouseScroll = [this](Event* event) { .. }
getEventDispatcher()->addEventListenerWithSceneGraphPriority(listener0, this);
```

**Comment by stevetranby** [comment]
If you want to submit a patch inside the experimental namespace they might merge into 3.9+, otherwise feel free to create a PR for v4.x

**Comment by pandamicro** [comment]
> I know you wanted to make it easy for Mouse users by just having 1 event listener for both, but that shit breaks down when you want to code for middle-click and right-click

Exactly, that's a desktop user's need, but what I want to clarify is that most of our user only targets on mobile platforms for deployment, not windows desktop. And I agree with you that it's annoying for users who targets on both desktop and mobile windows deployment.

I also like the Apple solution, it could be a solution to make everyone happy

**Comment by seivan** [comment]
@stevetranby That code won't work if you're abstracting clicks/touch. _Unless_ there is "data" in the Event that says this is a mouse click (doesn't need to say left, since only left works)
Then that might work

e.g.

``` cpp
struct TapEvent {

    enum class EventState: int {
     STARTED,
     CHANGED,
     COMPLETED,
     CANCELLED,
};
    enum class MouseButton: int {
        NONE = -1,
        LEFT,
        RIGHT,
        MIDDLE,
        BUTTON4,
        BUTTON5,
        BUTTON6,
        BUTTON7,
        BUTTON8,
    };

public:
    TapEvent(cocos2d::Touch &touch, EventState state) : button(MouseButton::NONE) {
        this->touch = &touch;
        this->touch->retain();
        this->state = state;
    }

    TapEvent(cocos2d::EventMouse &mouse, EventState state) : _isMouse(false) {
        this->mouse = &mouse;
        this->mouse->retain();
        this->state = state;
        this->button = MouseButton(mouse.getMouseButton());
    }

    TapEvent() {
        this->mouse->release();
        this->touch->release();
        this->mouse = nil;
        this->touch = nil;
    }

private: const bool _isMouse = false;
public: virtual bool isMouse(void) const { return this->_isMouse; }


    CC_SYNTHESIZE_READONLY(cocos2d::Touch, *touch, Touch);
    CC_SYNTHESIZE_READONLY(cocos2d::EventMouse, *mouse, Mouse);
    CC_SYNTHESIZE_READONLY(EventState, state, State);
    CC_SYNTHESIZE_READONLY(MouseButton, button, Button);

```

Would something like that be approved for 3.9? I genuinely don't want to sit on my own 4.0 fork if it won't be ready soon, but for 3.9 I'd love to rewrite the touch handling. I think it's pretty bad, borders to broken right now. 

 Would you accept a PR for 3.9 that makes touch/click handling agnostic - while keeping backwards compatibility? 
It would use _one_ listener for all, but it would send "data" (Event object) on the touch/click details, such as if it's a "long-press" or "middle-click"

Though I argue _again_ (heh), that Touch listener shouldn't even exist or compile for Desktop based targets. It doesn't make it easier, it just makes it more confusing. 
By doing that you essentially make Cocos2d harder to make real use for Desktop games, say an RTS that should work on mobile and desktop.

Every other user will experience the same problem with two events when they try that.

**Comment by stevetranby** [comment]
Sorry, I now realize that you are correct, thanks for making me think about it further.

We currently support left-drag RTS selection on Win32/Mac/Linux, but as you've made me realize we force the user interaction based on platform, not input device.

So, yes your ideas and code are great, as we've discussed. If you want any help or want me to test it out I'll gladly do so either pulling from your repo, or however. Maybe even a test project repo? 

Any PR for 3.9 will probably be accept if it doesn't change the behavior of anyone's game written for 3.x so if you're just adding information into the touch events it should be approved? ... but I can't speak for them, so confirm with pandamicro, or post on the dev section of the forums to get more devs looking at it if you haven't already.

**Comment by seivan** [comment]
I am not even sure why this is a "tech discussion" and not a bug. You're dispatching a touch event on a mouse click. That's just poor design through and through, it doesn't make things easier, as it executes both touch code as well as mouse code on Windows targets <- bug. 

If you want to be a cross platform dev tool, you have to be it full through and support Windows desktop without breaking it just because Cocos2d-x is popular on Android. 

Unless I am missing on how to differentiate between Windows Phone and Windows desktop.
Win32 is just Windows 8.

**Comment by stevetranby** [comment]
It is definitely NOT a bug. It is a design choice. Treat everything as touches. Offer the mouse listener and EventMouse only for those who want to try and customize the experience when using a mouse. Touch is the standard method of positional input. 

v4 can break this design choice. 
v3.9+ will probably only accept a PR that augments, not change this design choice.

You can have different build for WinPhone and Win32.

Use the `CC_PLATFORM_[name]` defines

[name] => { IOS, ANDROID, WP8, WINRT, WIN32, MAC, LINUX }

```
// desktops 
#if ((CC_TARGET_PLATFORM == CC_PLATFORM_WIN32) || (CC_TARGET_PLATFORM == CC_PLATFORM_MAC) || (CC_TARGET_PLATFORM == CC_PLATFORM_LINUX))
  // handle desktop specific code path
#endif
// mobile (phones/tablets)
#if ((CC_TARGET_PLATFORM == CC_PLATFORM_WP8) || (CC_TARGET_PLATFORM == CC_PLATFORM_WINRT) || (CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID)  || (CC_TARGET_PLATFORM == CC_PLATFORM_IOS))
  // code executes on mobile platforms only
#endif
```

WP8 covers all windows phones (8, 8.1, 10)
WINRT covers windows store (metro) apps (8, 8.1, 10)
WIN32 covers all windows desktop apps

Note that some tablets support both touch and mouse connected at the same time which is why I recommended a feature be added for detecting supported and currently connected input methods. That way you can code for touch and then add in support for differences in mouse input when a mouse is connected in the same way controllers can be detected.

**Comment by seivan** [comment]
Thanks for the explaining what macros are available and what they target, it wasn't so obvious I appreciate that.
I wonder if it's valid though, on Windows you compile once for all platforms. How does the macro work there? 

In the quest of making things _easy_ you make it harder by adding so many conditions to avoid _DOUBLE_ events being triggered.

It's actually much easier to tell the user to call this->jump() in two eventListeners than to have touch triggers being fired on mouse click. This makes the API more obvious.

Anyway, I've come as far as I can in trying to convince in you in dropping the touch listener for mouse.
Instead, I'll start focusing on a rewrite for an agnostic  TapEvent class that wraps both :) 

Thank you for the discussion, I appreciate it wasn't shut down completely.

**Comment by stevetranby** [comment]
You don't have to convince me (and again I'm just a game dev using the engine) and that's why I agree with your improvements to be added, I'm just trying to explain why they aren't there now and why you might find that they don't merge your additions until v4.0 and why you might have to add in a small amount of code for your needs to better detect capabilities to support mouse more extensively. 

If you don't start a forum discussion topic I'll start one so we can pull in the actual engine developers so they're aware of the issue(s). I don't think as many read through the github issues, nor as often.

The universal Windows package I'm not entirely sure as we're building for WinPhone/RT separately from Win32. I'll have to look into the universal build system to determine how to distinguish. If there's not a way currently then improving the device connected availability and support configuration system can definitely be merged into 3.8 or 3.9+. I would imagine the Windows build frameworks (and Visual Studio 2015) allow one to determine at runtime whether or not the platform is a phone/mini-tablet vs laptop/hybrid-tablet/desktop though if there really is only one executable run on all platforms.
